### PR TITLE
Computed node class

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds computed node classes.

@mitchellh @armon

### CODE REVIEW

File: nomad/structs/node_class.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	case "ID", "Name", "Links": // Uniquely identifying
+		return false, nil
+	case "Drain", "Status", "StatusDescription": // Set by server
+		return false, nil
+	case "ComputedClass": // Part of computed node class
+		return false, nil
+	case "CreateIndex", "ModifyIndex": // Raft indexes
+		return false, nil
+	case "Resources", "Reserved": // Doesn't effect placement capability
+		return false, nil
```
Comment (mitchellh) [inline nomad/structs/node_class.go]: I like the line splitting here for comment readability. :+1:

File: client/fingerprint/env_aws.go [hunk 2]
```diff
@@ -114,6 +115,17 @@ func (f *EnvAWSFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
 		"public-ipv4",
 		"placement/availability-zone",
 	}
+
+	// Keys that should be marked as unique
+	unique := map[string]struct{}{
```
Comment (armon) [inline client/fingerprint/env_aws.go]: Make the above map a `map[string]bool` to not accidentally mess up new keys. Add more docs as to why.

File: client/fingerprint/env_gce.go [hunk 3]
```diff
@@ -140,6 +141,12 @@ func (f *EnvGCEFingerprint) Fingerprint(cfg *config.Config, node *structs.Node)
 		"scheduling/automatic-restart",
 		"scheduling/on-host-maintenance",
 	}
+
+	// Keys that should be marked as unique
+	unique := map[string]struct{}{
```
Comment (armon) [inline client/fingerprint/env_gce.go]: Same issue

File: nomad/structs/node_class.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	default:
+		return true, nil
```
Comment (armon) [inline nomad/structs/node_class.go]: It may be safer to white list and move this to default false and explicitly allow certain fields.

**Comment by armon** [comment]
LGTM otherwise

**Pull request merged**